src/main/client/batch_read.c

```diff
@@ -0,0 +1,299 @@
+
+        Py_XINCREF(py_key);
+        if (!PyTuple_Check(py_key)) {
+            as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                            "key should be an aerospike key tuple");
+            Py_XDECREF(py_key);
+            goto CLEANUP2;
+        }
+
+        pyobject_to_key(&err, py_key, tmp_key);
+        if (err.code != AEROSPIKE_OK) {
+            as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                            "failed to convert key at index: %d", i);
+            goto CLEANUP2;
+        }
+
+        processed_key_count++;
+    }
+
+    as_batch batch;
+    as_batch_init(&batch, processed_key_count);
+    memcpy(batch.keys.entries, tmp_keys.list,
+           sizeof(as_key) * processed_key_count);
+
+    as_policy_batch policy_batch;
+    as_policy_batch *policy_batch_p = NULL;
+
+    // For expressions conversion.
+    as_exp batch_exp_list;
+    as_exp *batch_exp_list_p = NULL;
+
+    if (py_policy_batch) {
+        if (pyobject_to_policy_batch(
+                self, &err, py_policy_batch, &policy_batch, &policy_batch_p,
+                &self->as->config.policies.batch, &batch_exp_list,
+                &batch_exp_list_p) != AEROSPIKE_OK) {
+            goto CLEANUP3;
+        }
+    }
+
+    // import batch_records helper
+    PyObject *br_module = NULL;
+    PyObject *sys_modules = PyImport_GetModuleDict();
+
+    Py_INCREF(sys_modules);
+    if (PyMapping_HasKeyString(sys_modules,
+                               "aerospike_helpers.batch.records")) {
+        br_module = PyMapping_GetItemString(sys_modules,
+                                            "aerospike_helpers.batch.records");
+    }
+    else {
+        br_module = PyImport_ImportModule("aerospike_helpers.batch.records");
+    }
+    Py_DECREF(sys_modules);
+
+    if (!br_module) {
+        as_error_update(&err, AEROSPIKE_ERR_CLIENT,
+                        "Unable to load batch_records module");
+        goto CLEANUP3;
+    }
+
+    PyObject *obj_name = PyUnicode_FromString("BatchRecords");
+    PyObject *res_list = PyList_New(0);
+    PyObject *br_instance =
+        PyObject_CallMethodObjArgs(br_module, obj_name, res_list, NULL);
+
+    Py_DECREF(obj_name);
+    Py_DECREF(res_list);
+
+    if (!br_instance) {
+        as_error_update(&err, AEROSPIKE_ERR_CLIENT,
+                        "Unable to instance BatchRecords");
+        goto CLEANUP4;
+    }
+
+    // Create and initialize callback user-data
+    LocalData data;
+    // Used to decode record bins
+    data.client = self;
+    // Used to append BatchRecord instances to the BatchRecords object in this function
+    data.py_results = PyObject_GetAttrString(br_instance, "batch_records");
+    // Used to create a new BatchRecord instance in the callback function
+    data.batch_records_module = br_module;
+    data.func_name = PyUnicode_FromString("BatchRecord");
+    data.checking_if_records_exist = false;
+
+    Py_ssize_t bin_count = 0;
+    char **filter_bins = NULL;
```

Comment:
I think this should be `const char **` since it is used with PyUnicode_AsUTF8 which returns `const * char`. Do you get a compiler warning for this or are the macros still covering it up?

Reply:
Fixed.

Reply:
yeah I had to check TBH https://cdecl.org/?q=declare+x+as+pointer+to+pointer+to+const+char